File: code/ex91.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n = s.length();
        vector<int> dp(n + 1);
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = max(0, i - 2); j < i; j++) {
                string ss = s.substr(j, i - j);
                if (valid(ss))    dp[i] += dp[j];
            }
        }
        return dp.back();
    }
    
private:
    bool valid(string s) {
        int n = s.length();
        if (n == 1)    return s > "0" && s <= "9";
        else    return s >= "10" && s <= "26";
    }
};

File: code/ex91.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        
        int[] dp = new int[n];
        dp[0] = s.charAt(0) == '0' ? 0 : 1;
        
        for (int i = 1; i < n; i++) {
            for (int j = Math.max(0, i - 1); j <= i; j++) {
                if (s.charAt(j) == '0')    continue;
                int num = Integer.parseInt(s.substring(j, i + 1));
                if (!(num >= 1 && num <= 26))    continue;
                dp[i] += j > 0 ? dp[j - 1] : 1;
            }
        }
        
        return dp[n - 1];
    }
}